File: swissknife/licensing.py
from typing import Any, Dict, List

from .graph_client import GraphClient


def list_skus(client: GraphClient) -> List[Dict[str, Any]]:
    """
    Список подписок (SKU) тенанта.
    GET /subscribedSkus
    Требуются Directory.Read.All / Organization.Read.All (Application).
    """
    result = client.get("/subscribedSkus")
    return result.get("value", [])


def assign_licenses(
    client: GraphClient,
    user: str,
    add_sku_ids: List[str],
    remove_sku_ids: List[str],
) -> Dict[str, Any]:
    """
    Выдать / снять лицензии пользователя.
    POST /users/{user}/assignLicense

    add_sku_ids  — список GUID SKU, которые нужно ДОБАВИТЬ.
    remove_sku_ids — список GUID SKU, которые нужно СНЯТЬ.
    """
    body = {
        "addLicenses": [{"skuId": sku} for sku in add_sku_ids],
        "removeLicenses": remove_sku_ids,
    }

    path = f"/users/{user}/assignLicense"
    return client.post(path, json=body)

File: app/internal/services/licensing.go
package services

import (
	"encoding/json"
	"net/url"
	"strconv"

	"swissknife-app/internal/session"
)

type LicensingService struct {
	s *session.Session
}

func NewLicensingService(s *session.Session) *LicensingService { return &LicensingService{s: s} }

func (l *LicensingService) Skus() ([]json.RawMessage, error) {
	c, err := l.s.Client()
	if err != nil {
		return nil, err
	}
	return c.ListAll(l.s.Ctx(), "/subscribedSkus", nil, 0)
}

func (l *LicensingService) Assign(user string, addSkuIDs, removeSkuIDs []string) (json.RawMessage, error) {
	if err := l.s.GuardWrite(); err != nil {
		return nil, err
	}
	c, err := l.s.Client()
	if err != nil {
		return nil, err
	}
	add := make([]map[string]any, 0, len(addSkuIDs))
	for _, id := range addSkuIDs {
		add = append(add, map[string]any{"skuId": id})
	}
	if removeSkuIDs == nil {
		removeSkuIDs = []string{}
	}
	body := map[string]any{"addLicenses": add, "removeLicenses": removeSkuIDs}
	var out json.RawMessage
	err = c.Post(l.s.Ctx(), "/users/"+url.PathEscape(user)+"/assignLicense", body, &out)
	l.s.Record("licensing.assign", user,
		"add="+strconv.Itoa(len(addSkuIDs))+" remove="+strconv.Itoa(len(removeSkuIDs)), err)
	return out, err
}
